lightning/src/ln/outbound_payment.rs

```diff
@@ -896,6 +954,119 @@ impl OutboundPayments {
 		Ok(())
 	}
 
+	#[cfg(async_payments)]
+	pub(super) fn static_invoice_received<ES: Deref>(
+		&self, invoice: &StaticInvoice, payment_id: PaymentId, features: Bolt12InvoiceFeatures,
+		best_block_height: u32, entropy_source: ES,
+		pending_events: &Mutex<VecDeque<(events::Event, Option<EventCompletionAction>)>>
+	) -> Result<[u8; 32], Bolt12PaymentError> where ES::Target: EntropySource {
+		macro_rules! abandon_with_entry {
+			($payment: expr, $reason: expr) => {
+				$payment.get_mut().mark_abandoned($reason);
+				if let PendingOutboundPayment::Abandoned { reason, .. } = $payment.get() {
+					if $payment.get().remaining_parts() == 0 {
+						pending_events.lock().unwrap().push_back((events::Event::PaymentFailed {
+							payment_id,
+							payment_hash: None,
+							reason: *reason,
+						}, None));
+						$payment.remove();
+					}
+				}
+			}
+		}
+
+		match self.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+			hash_map::Entry::Occupied(mut entry) => match entry.get() {
+				PendingOutboundPayment::AwaitingInvoice { retry_strategy, invoice_request, .. } => {
+					let invreq = invoice_request.as_ref().ok_or(Bolt12PaymentError::UnexpectedInvoice)?;
+					if !invoice.from_same_offer(invreq) {
+						return Err(Bolt12PaymentError::UnexpectedInvoice)
+					}
+					if invoice.invoice_features().requires_unknown_bits_from(&features) {
+						abandon_with_entry!(entry, PaymentFailureReason::UnknownRequiredFeatures);
+						return Err(Bolt12PaymentError::UnknownRequiredFeatures)
+					}
+					let amount_msat = match InvoiceBuilder::<DerivedSigningPubkey>::amount_msats(&invreq) {
+						Ok(amt) => amt,
+						Err(_) => {
+							debug_assert!(false, "LDK requires an invreq amount for amount-less offers");
+							abandon_with_entry!(entry, PaymentFailureReason::UnexpectedError);
+							return Err(Bolt12PaymentError::UnknownRequiredFeatures)
+						}
+					};
+					let keysend_preimage = PaymentPreimage(entropy_source.get_secure_random_bytes());
+					let payment_hash = PaymentHash(Sha256::hash(&keysend_preimage.0).to_byte_array());
+					let payment_release_secret = entropy_source.get_secure_random_bytes();
+					let pay_params = PaymentParameters::from_static_invoice(invoice);
+					let mut route_params = RouteParameters::from_payment_params_and_value(pay_params, amount_msat);
+
+					if let Err(()) = onion_utils::set_max_path_length(
+						&mut route_params, &RecipientOnionFields::spontaneous_empty(), Some(keysend_preimage),
+						best_block_height
+					) {
+						abandon_with_entry!(entry, PaymentFailureReason::UnexpectedError);
```

Comment:
Updated this. Btw, I got these error conversions from here:
https://github.com/lightningdevkit/rust-lightning/blob/82b3f620f587abb9561482f9d332218d8ba528f3/lightning/src/ln/outbound_payment.rs#L854-L859
So I think that callsite could use an update too. I looked into it briefly, but it's not entirely straightforward since the `RetryablePayment::DuplicatePayment` case is currently unreachable, which may warrant swapping for a different error.... so I left the linked code as-is for now.